Test onedata objectstore with new caching

bwalkowi opened the pull request:
This PR complements this one: https://github.com/galaxyproject/galaxy/pull/18174

We have tested the deduplicated Onedada object store code and confirmed that it works.

We also took the occasion to include some improvements to our libs. They make the Onedata clients resistant to failures of data providers and improve their performance.

NOTE: the OnedataFileRestClient logs on the debug level, at least one log per request. Let us know if it's okay or maybe we should adjust the Galaxy's logger config so that these logs are ignored.

NOTE: while working on this, we have stumbled upon major performance problems which can also cause ambiguous data access errors, as described in https://github.com/galaxyproject/galaxy/issues/18369. This is why we are proposing to change the implementation of `_get_total_matches_count` to use scandir instead of glob in this PR.

[x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

bwalkowi committed:
Replace demo.onedata.org with datahub.egi.eu

bwalkowi committed:
Update OnedataObjectStore to new onedatafilerestclient API

bwalkowi committed:
Update onedatafilerestclient requirement

lopiola commented:
Proponuję docelowy opis PRa:

This PR complements this one: https://github.com/galaxyproject/galaxy/pull/18174

We have tested the deduplicated Onedada object store code and confirmed that it works.

We also took the occasion to include some improvements to our libs. They make the Onedata clients resistant to failures of data providers and improve their performance. 

**NOTE**: the `OnedataFileRestClient` logs on the `debug` level, at least one log per request. Let us know if it's okay or maybe we should adjust the Galaxy's logger config so that these logs are ignored.

[x]  This is a refactoring of components with existing test coverage.

lopiola commented:
Najładniej by było chyba zrobić PRa do brancha jmchiltona, żeby jak sobie to wciągnie mogło wejść wszystko na raz z jego deduplikacją?

lopiola commented on lib/galaxy/config/__init__.py:
```diff
@@ -116,6 +116,14 @@
+        "onedatafilerestclient.http_client": {
+            "level": "INFO",
+            "qualname": "onedatafilerestclient.http_client",
+        },
```
zgaduję, że nie będą tego potrzebować więc póki co bym usunął

w opisie PRa zaproponowałem notkę dla nich do wyjaśnienia

lopiola commented on lib/galaxy/config/sample/object_store_conf.sample.yml:
```diff
@@ -223,9 +223,9 @@ auth:
   access_token: ...
 connection:
-  # the domain of the Onezone service (e.g. "demo.onedata.org"), or its IP address for
+  # the domain of the Onezone service (e.g. datahub.egi.eu), or its IP address for
```
thx

lopiola commented on lib/galaxy/dependencies/dev-requirements.txt:
```diff
@@ -79,7 +79,9 @@ myst-parser==3.0.1 ; python_version >= "3.8" and python_version < "3.13"
-onedatafilerestclient==21.2.5rc1 ; python_version >= "3.8" and python_version < "3.13"
+# onedatafilerestclient==21.2.5.1 ; python_version >= "3.8" and python_version < "3.13"
+../onedatafilerestclient ; python_version >= "3.8" and python_version < "3.13"
+../onedatarestfs ; python_version >= "3.8" and python_version < "3.13"
```
to na koniec do ogarnięcia

lopiola commented on lib/galaxy/objectstore/onedata.py:
```diff
@@ -7,8 +7,9 @@
 try:
-    from onedatafilerestclient import (
-        OnedataFileRESTClient,
+    from onedatafilerestclient import OnedataFileRESTClient
```
wygląda na to że nie pykło z datahubem - dyskusja na Slaku

mvdbeek committed:
Don't copy from_work_dir outputs to purged path in shell fragment

mvdbeek committed:
Don't include purged outputs in from_work_dir outputs

mvdbeek committed:
Raise appropriate exception if require path parameter is missing

mvdbeek committed:
Don't push purged dataset contents to object store

This in particular needs a lot of new tests.
We will also need to actively purge datasets in the model store import
code, since users might have purged datasets while the job ran.
Again, more tests needed.

mvdbeek committed:
Add tool that produces all types of outputs

mvdbeek committed:
Don't fail _finish_or_resubmit if tool_stdout / tool_stderr not written

These are only written if the command actually ran, so we'd fail here
for instance if the outputs are deleted before the job had a chance to
run.

mvdbeek committed:
Fix up test cases

Add a sleep so we can delay output purging until command line is
templated. If we don't sleep we generate a command line where the
output path is just `''`. That needs another fix!

mvdbeek committed:
Fix handling of collecting discovered but purged outputs

mvdbeek committed:
Elevate external metadata setting to warning

This is a tricky one. There could be systematic issues here that an
admin would want to fix, but it could also just be the case that a user
forced a wrong datatype. Ideally we'd probably tag this as job execution
issue ? It's not very different from a job error IMO.

mvdbeek committed:
Fix metadata setting on purged outputs

mvdbeek committed:
Extend test to make sure disk path remains purged

mvdbeek committed:
Set file_size and total_size to 0 for discovered but purged datasets

davelopez committed:
Fix anonymous user job retrieval logic

For anonymous users, if a history_id is provided, do not override with the current session history and instead rely on the history accessibility

davelopez committed:
Filter jobs by session instead of using the session current history for anonymous

Co-authored-by: mvdbeek <m.vandenbeek@gmail.com>

mvdbeek committed:
Purge dataset if finishing job for purged output

mvdbeek committed:
Do not overwrite purged or deleted states when importing datasets from job

mvdbeek committed:
Delete purged outputs from object store when importing dataset for purged output

mvdbeek committed:
Fix up tests

davelopez committed:
Require session to list jobs instead of returning empty job list for anonymous

Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

jdavcs committed:
Fix check for anonymous

martenson committed:
Merge pull request #18358 from davelopez/24.0_fix_condition_anonymous_jobs

[24.0] Fix anonymous user job retrieval logic

jdavcs committed:
Merge pull request #18364 from jdavcs/241_bug_reverse_bool

[24.1] Fix check for anonymous

jdavcs committed:
Merge branch 'release_24.0' into release_24.1

jdavcs committed:
Fix mypy after merge

jdavcs committed:
Merge pull request #18365 from jdavcs/241_merge_240_2

[24.1] Merge 24.0 into 24.1

itisAliRH committed:
✨: add `Link to workflow to` WorkflowActionsExtend

itisAliRH committed:
🎨: replace `workflow external link` icon with `faFileImport` in `WorkflowIndicators`

bwalkowi committed:
Use only single filterdir in _get_total_matches_count

mvdbeek committed:
Merge pull request #18342 from mvdbeek/fix_copying_purged_files

[24.0] Do not copy purged outputs to object store

mvdbeek committed:
Merge branch 'release_24.0' into release_24.1

itisAliRH committed:
✨: add `getFullAppUrl` utility function

itisAliRH committed:
🔨: replace direct URL construction with the `getFullAppUrl` utility function in `WorkflowEmbed`, `SharingPage`, `WorkflowInformation` and `WorkflowActionsExtend`

itisAliRH committed:
🐛: remove unnecessary condition for anonymous user in `WorkflowActionsExtend`

itisAliRH committed:
🚨: add unit test for `getFullAppUrl` utility function

ahmedhamidawan committed:
Merge pull request #18370 from itisAliRH/workflow-card-publish-link-copy

[24.1] Add copy link to published workflow in `WorkflowCard`

bwalkowi committed:
Use basic namespace when counting files with scandir

Co-authored-by: David López <46503462+davelopez@users.noreply.github.com>
